mm2src/coins/coin_balance.rs

```diff
@@ -0,0 +1,452 @@
+    /// Please note the function can return zero balance even if the address has been used before.
+    async fn check_address_balance(
+        &self,
+        address: &Self::Address,
+        checker: &Self::HDAddressChecker,
+    ) -> BalanceResult<AddressBalanceStatus<CoinBalance>> {
+        if !checker.is_address_used(address).await? {
+            return Ok(AddressBalanceStatus::NotUsed);
+        }
+        let balance = self.address_balance(address).await?;
+        Ok(AddressBalanceStatus::Used(balance))
+    }
+}
+
+#[async_trait]
+pub trait HDAddressBalanceChecker: Sync {
+    type Address;
+
+    async fn is_address_used(&self, address: &Self::Address) -> BalanceResult<bool>;
+}
+
+pub enum AddressBalanceStatus<Balance> {
+    Used(Balance),
+    NotUsed,
+}
+
+#[async_trait]
+pub trait AddressBalanceOps {
+    type Address: Sync;
+
+    async fn address_balance(&self, address: &Self::Address) -> BalanceResult<CoinBalance>;
+}
+
+#[async_trait]
+pub trait HDWalletBalanceRpcOps: HDWalletCoinOps {
+    async fn hd_account_balance_rpc(
+        &self,
+        params: HDAccountBalanceParams,
+    ) -> MmResult<HDAccountBalanceResponse, HDAccountBalanceRpcError>;
+
+    async fn check_hd_account_balance_rpc(
+        &self,
+        params: CheckHDAccountBalanceParams,
+    ) -> MmResult<CheckHDAccountBalanceResponse, HDAccountBalanceRpcError>;
+}
+
+pub async fn hd_account_balance(
+    ctx: MmArc,
+    req: HDAccountBalanceRequest,
+) -> MmResult<HDAccountBalanceResponse, HDAccountBalanceRpcError> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    match coin {
+        MmCoinEnum::UtxoCoin(utxo) => utxo.hd_account_balance_rpc(req.params).await,
+        MmCoinEnum::QtumCoin(qtum) => qtum.hd_account_balance_rpc(req.params).await,
+        _ => MmError::err(HDAccountBalanceRpcError::ExpectedHDWalletDerivationMethod {
+            coin: coin.ticker().to_owned(),
+        }),
+    }
+}
+
+pub async fn check_hd_account_balance(
+    ctx: MmArc,
+    req: CheckHDAccountBalanceRequest,
+) -> MmResult<CheckHDAccountBalanceResponse, HDAccountBalanceRpcError> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    match coin {
+        MmCoinEnum::UtxoCoin(utxo) => utxo.check_hd_account_balance_rpc(req.params).await,
+        MmCoinEnum::QtumCoin(qtum) => qtum.check_hd_account_balance_rpc(req.params).await,
+        _ => MmError::err(HDAccountBalanceRpcError::ExpectedHDWalletDerivationMethod {
+            coin: coin.ticker().to_owned(),
+        }),
+    }
+}
+
+pub mod common_impl {
+    use super::*;
+    use crate::hd_wallet::HDAddress;
+    use common::calc_total_pages;
+
+    pub(crate) async fn enable_hd_wallet_balance<Coin, Address, HDWallet, HDAccount, HDAccountChecker>(
+        coin: &Coin,
+        hd_wallet: &HDWallet,
+    ) -> BalanceResult<HDWalletBalance>
+    where
+        Coin: HDWalletCoinOps<Address = Address, HDWallet = HDWallet, HDAccount = HDAccount>
```

Comment:
I guess this bound can be simplified by requiring `HDWalletCoinOps: HDWalletBalanceOps`.

Reply:
I think there can be some coins that don't implement `HDWalletBalanceOps` but implement `HDWalletCoinOps`. For now, we can make `HDWalletCoinOps` the `HDWalletBalanceOps` trait dependent.

Reply:
I would even do the following:
`HDWalletBalanceOps : HDWalletCoinOps`. `HDWalletCoinOps` doesn't contain methods related to the balance checking, but `HDWalletBalanceOps` is related to the HD wallet functionality.

Reply:
I jumped to the conclusion. It will be complicated and lead to more boilerplate code because both `AddressBalanceOps` and `HDWalletCoinOps` have `type Address`. Currently, rust doesn't allow to declare `AddressBalanceOps::Address == HDWalletCoinOps::Address`. In the case when `HDWalletBalanceOps : HDWalletCoinOps` we should do the following:
```rust
trait HDWalletBalanceOps<Address>:
    HDWalletCoinOps<Address=Address>
    + AddressBalanceOps<Address=Address> {
}
```
Honestly, I would like to avoid such dependencies.
What do you think about adding a trait alias?
```rust
trait HDWalletCoinAndBalanceOps<HDWallet, HDAccount, HDAddressChecker, Address>:
    HDWalletCoinOps<HDWallet=HDWallet, HDAccount=HDAccount, Address=Address>
    + HDWalletBalanceOps<HDWallet=HDWallet, HDAccount=HDAccount, HDAddressChecker=HDAddressChecker, Address=Address>
    + HDWalletBalanceOps<HDWallet=HDWallet, HDAccount=HDAccount, HDAddressChecker=HDAddressChecker, Address=Address>
    + AddressBalanceOps<Address=Address> {
}
```

Reply:
I've added the trait alias [here](https://github.com/KomodoPlatform/atomicDEX-API/commit/e0f1795b788efa94b99ed3af213241aa8d6be3b6). If it's not ok, I can roll it back easily.

Reply:
I have noticed that some traits have similarly named associated types, which can be actually inherited and then resolved by the compiler. `AddressBalanceOps` looks excessive as it contains only one method. I think `address_balance` fits `HDWalletBalanceOps` well. I have tried to implement this and a bit of other refactoring in https://github.com/KomodoPlatform/atomicDEX-API/commit/79d544932dabc69f945544097ae67e1fecb58050. Could you check it, please? 
If you accept this, I think you may find a way to simplify some of the trait bounds and reduce the number of generics even more 🙂

Reply:
Thank you for the huge help! I've finished your refactoring https://github.com/KomodoPlatform/atomicDEX-API/pull/1211, it resulted in a significant reduction in the number of generic types and trait bounds.